File: ios/runanywhere-sdks-latest/sdk/runanywhere-web/packages/core/src/services/HTTPService.ts
/**
 * HTTPService.ts
 *
 * Core HTTP service for the RunAnywhere Web SDK.
 * Ported from sdk/runanywhere-react-native/packages/core/src/services/Network/HTTPService.ts
 * Adapted for browser (uses native fetch, AbortController, setTimeout globals).
 *
 * Responsibilities:
 * - Centralized HTTP transport layer for telemetry, device registration, etc.
 * - Environment-aware routing (Supabase for dev, Railway for prod/staging)
 * - Automatic header management (API key, auth tokens, SDK metadata)
 */

import { SDKLogger } from '../Foundation/SDKLogger';
import { SDKError, SDKErrorCode } from '../Foundation/ErrorTypes';
import { SDKEnvironment } from '../types/enums';

const logger = new SDKLogger('HTTPService');

const SDK_CLIENT = 'RunAnywhereSDK';
const SDK_PLATFORM = 'web';
const SDK_VERSION = '0.1.0-beta.8';
const DEFAULT_TIMEOUT_MS = 30000;
const DEVICE_ID_KEY = 'rac_device_id';
const TELEMETRY_TABLE = 'rest/v1/telemetry_events';

// NOTE: TELEMETRY_COLUMNS filter removed — all telemetry now flows through
// the C++ telemetry manager via AnalyticsEmitter → rac_analytics_emit_*().

/**
 * HTTP Service Configuration for non-dev environments.
 */
export interface HTTPServiceConfig {
  /** Base URL for API requests */
  baseURL: string;
  /** API key for authentication */
  apiKey: string;
  /** SDK environment */
  environment: SDKEnvironment;
  /** Request timeout in milliseconds */
  timeoutMs?: number;
}

/**
 * Development (Supabase) Configuration
 */
export interface DevModeConfig {
  /** Supabase project URL */
  supabaseURL: string;
  /** Supabase anon key */
  supabaseKey: string;
}

/**
 * HTTPService - Centralized HTTP transport layer for the Web SDK.
 *
 * Environment-aware routing:
 * - Development: Supabase credentials compiled into WASM (rac_dev_config_*)
 * - Staging/Production: Railway backend with API key
 */
export class HTTPService {
  private static _instance: HTTPService | null = null;

  static get shared(): HTTPService {
    if (!HTTPService._instance) {
      HTTPService._instance = new HTTPService();
    }
    return HTTPService._instance;
  }

  private baseURL: string = '';
  private apiKey: string = '';
  private environment: SDKEnvironment = SDKEnvironment.Production;
  private accessToken: string | null = null;
  private timeoutMs: number = DEFAULT_TIMEOUT_MS;

  private supabaseURL: string = '';
  private supabaseKey: string = '';

  private constructor() {}

  private get defaultHeaders(): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'X-SDK-Client': SDK_CLIENT,
      'X-SDK-Version': SDK_VERSION,
      'X-Platform': SDK_PLATFORM,
    };
  }

  // ---------------------------------------------------------------------------
  // Telemetry helpers
  // ---------------------------------------------------------------------------

  /**
   * @deprecated Use `AnalyticsEmitter` instead.  All telemetry now routes
   * through the C++ telemetry manager via `rac_analytics_emit_*()`.
   * Kept as a fallback for edge cases where C++ is unavailable.
   */
  postTelemetryEvent(partialPayload: Record<string, unknown>): void {
    if (!this.isConfigured) return;

    const payload: Record<string, unknown> = {
      sdk_event_id: crypto.randomUUID(),
      event_timestamp: new Date().toISOString(),
      created_at: new Date().toISOString(),
      device_id: this.getOrCreateDeviceId(),
      platform: SDK_PLATFORM,
      sdk_version: SDK_VERSION,
      ...partialPayload,
    };

    const url = this.buildFullURL(TELEMETRY_TABLE);
    const headers = this.buildHeaders(false);
    this.executeRequest('POST', url, headers, [payload]).catch(() => { /* silent */ });
  }

  /**
   * Returns the persistent device UUID, creating one if it doesn't exist.
   * Mirrors getOrCreateDeviceId() in TelemetryService.ts.
   */
  getOrCreateDeviceId(): string {
    try {
      const existing = localStorage.getItem(DEVICE_ID_KEY);
      if (existing) return existing;
      const id = crypto.randomUUID();
      localStorage.setItem(DEVICE_ID_KEY, id);
      return id;
    } catch {
      return crypto.randomUUID();
    }
  }

  // ---------------------------------------------------------------------------
  // Configuration
  // ---------------------------------------------------------------------------

  configure(config: HTTPServiceConfig): void {
    this.baseURL = config.baseURL;
    this.apiKey = config.apiKey;
    this.environment = config.environment;
    this.timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    logger.info(`Configured for ${config.environment} environment`);
  }

  /**
   * Configure development mode using Supabase credentials.
   * Called during WASM init using credentials read from rac_dev_config_*.
   */
  configureDev(config: DevModeConfig): void {
    this.supabaseURL = config.supabaseURL;
    this.supabaseKey = config.supabaseKey;
    this.environment = SDKEnvironment.Development;
    logger.info('Development mode configured with Supabase');
  }

  setToken(token: string): void {
    this.accessToken = token;
  }

  clearToken(): void {
    this.accessToken = null;
  }

  get isConfigured(): boolean {
    if (this.environment === SDKEnvironment.Development) {
      return !!this.supabaseURL;
    }
    return !!this.baseURL && !!this.apiKey;
  }

  get currentBaseURL(): string {
    if (this.environment === SDKEnvironment.Development && this.supabaseURL) {
      return this.supabaseURL;
    }
    return this.baseURL;
  }

  // ---------------------------------------------------------------------------
  // HTTP Methods
  // ---------------------------------------------------------------------------

  async post<T = unknown, R = unknown>(path: string, data?: T): Promise<R> {
    let url = this.buildFullURL(path);
    const isDeviceReg = this.isDeviceRegistrationPath(path);
    const headers = this.buildHeaders(isDeviceReg);

    if (isDeviceReg && this.environment === SDKEnvironment.Development) {
      const separator = url.includes('?') ? '&' : '?';
      url = `${url}${separator}on_conflict=device_id`;
    }

    const response = await this.executeRequest('POST', url, headers, data);

    if (isDeviceReg && response.status === 409) {
      logger.info('Device already registered (409) — treating as success');
      return this.parseResponse<R>(response);
    }

    return this.handleResponse<R>(response, path);
  }

  async get<R = unknown>(path: string): Promise<R> {
    const url = this.buildFullURL(path);
    const headers = this.buildHeaders(false);
    const response = await this.executeRequest('GET', url, headers);
    return this.handleResponse<R>(response, path);
  }

  async delete<R = unknown>(path: string): Promise<R> {
    const url = this.buildFullURL(path);
    const headers = this.buildHeaders(false);
    const response = await this.executeRequest('DELETE', url, headers);
    return this.handleResponse<R>(response, path);
  }

  // ---------------------------------------------------------------------------
  // Private
  // ---------------------------------------------------------------------------

  private async executeRequest<T>(
    method: string,
    url: string,
    headers: Record<string, string>,
    data?: T,
  ): Promise<Response> {
    logger.debug(`${method} ${url}`);

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const options: RequestInit = { method, headers, signal: controller.signal };
      if (data !== undefined && method !== 'GET') {
        options.body = JSON.stringify(data);
      }
      return await fetch(url, options);
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private buildHeaders(isDeviceRegistration: boolean): Record<string, string> {
    const headers = { ...this.defaultHeaders };

    if (this.environment === SDKEnvironment.Development) {
      if (this.supabaseKey) {
        headers['apikey'] = this.supabaseKey;
        headers['Authorization'] = `Bearer ${this.supabaseKey}`;
        headers['Prefer'] = isDeviceRegistration ? 'resolution=merge-duplicates' : 'return=representation';
      }
    } else {
      const token = this.accessToken || this.apiKey;
      if (token) {
        headers['Authorization'] = `Bearer ${token}`;
      }
    }

    return headers;
  }

  private buildFullURL(path: string): string {
    if (path.startsWith('http://') || path.startsWith('https://')) {
      return path;
    }
    const base = this.currentBaseURL.replace(/\/$/, '');
    const endpoint = path.startsWith('/') ? path : `/${path}`;
    return `${base}${endpoint}`;
  }

  private isDeviceRegistrationPath(path: string): boolean {
    return path.includes('sdk_devices') || path.includes('devices/register') || path.includes('rest/v1/sdk_devices');
  }

  private async parseResponse<R>(response: Response): Promise<R> {
    const text = await response.text();
    if (!text) return {} as R;
    try {
      return JSON.parse(text) as R;
    } catch {
      return text as unknown as R;
    }
  }

  private async handleResponse<R>(response: Response, path: string): Promise<R> {
    if (response.ok) {
      return this.parseResponse<R>(response);
    }

    let errorMessage = `HTTP ${response.status}`;
    try {
      const errorData = (await response.json()) as Record<string, unknown>;
      errorMessage = (errorData.message as string) || (errorData.error as string) || errorMessage;
    } catch {
      // ignore
    }

    // Telemetry failures are non-critical — log at debug level only
    if (path.includes('telemetry')) {
      logger.debug(`HTTP ${response.status}: ${path}`);
    } else {
      logger.error(`HTTP ${response.status}: ${path}`);
    }
    throw this.mapHttpError(response.status, errorMessage);
  }

  private mapHttpError(status: number, message: string): SDKError {
    switch (status) {
      case 401:
      case 403:
        return new SDKError(SDKErrorCode.AuthenticationFailed, message);
      case 408:
      case 429:
        return new SDKError(SDKErrorCode.NetworkTimeout, message);
      default:
        return new SDKError(SDKErrorCode.NetworkError, `HTTP ${status}: ${message}`);
    }
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Data/Network/Services/HTTPService.swift
//
//  HTTPService.swift
//  RunAnywhere SDK
//
//  Core HTTP service implementation using URLSession.
//  All network logic is centralized here.
//

import CRACommons
import Foundation

/// HTTP Service - Core network implementation
/// Centralized HTTP transport layer using URLSession
public actor HTTPService: NetworkService {

    // MARK: - Singleton

    /// Shared HTTP service instance
    public static let shared = HTTPService()

    // MARK: - Configuration

    private var session: URLSession
    private var baseURL: URL?
    private var apiKey: String?
    private let logger = SDKLogger(category: "HTTPService")

    // MARK: - Initialization

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30.0
        config.httpAdditionalHeaders = Self.defaultHeaders
        self.session = URLSession(configuration: config)
    }

    private static var defaultHeaders: [String: String] {
        [
            "Content-Type": "application/json",
            "Accept": "application/json",
            "X-SDK-Client": "RunAnywhereSDK",
            "X-SDK-Version": SDKConstants.version,
            "X-Platform": SDKConstants.platform
        ]
    }

    // MARK: - Configuration

    /// Configure HTTP service with base URL and API key
    public func configure(baseURL: URL, apiKey: String) {
        self.baseURL = baseURL
        self.apiKey = apiKey

        // Update session with API key header
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30.0
        config.httpAdditionalHeaders = Self.defaultHeaders.merging([
            "apikey": apiKey,
            "Prefer": "return=representation"
        ]) { _, new in new }
        self.session = URLSession(configuration: config)

        logger.info("HTTP service configured with base URL: \(baseURL.host ?? "unknown")")
    }

    /// Configure with URL string
    public func configure(baseURL: String, apiKey: String) {
        guard let url = URL(string: baseURL) else {
            logger.error("Invalid base URL: \(baseURL)")
            return
        }
        configure(baseURL: url, apiKey: apiKey)
    }

    /// Check if HTTP is configured
    public var isConfigured: Bool {
        baseURL != nil
    }

    /// Current base URL
    public var currentBaseURL: URL? {
        baseURL
    }

    // MARK: - NetworkService Protocol

    /// POST request with raw Data body
    public func postRaw(
        _ path: String,
        _ payload: Data,
        requiresAuth: Bool
    ) async throws -> Data {
        // For Supabase device registration, use UPSERT (merge-duplicates) to handle existing devices
        // Supabase PostgREST requires both:
        // 1. The `Prefer: resolution=merge-duplicates` header
        // 2. The `?on_conflict=device_id` query parameter to specify the conflict column
        if path.contains(RAC_ENDPOINT_DEV_DEVICE_REGISTER) {
            // Add on_conflict query parameter to the path
            let upsertPath = path.contains("?") ? "\(path)&on_conflict=device_id" : "\(path)?on_conflict=device_id"
            return try await postRawWithHeaders(
                upsertPath,
                payload,
                requiresAuth: requiresAuth,
                additionalHeaders: ["Prefer": "resolution=merge-duplicates"]
            )
        }
        return try await postRawWithHeaders(path, payload, requiresAuth: requiresAuth)
    }

    /// POST request with raw Data body and optional additional headers
    private func postRawWithHeaders(
        _ path: String,
        _ payload: Data,
        requiresAuth: Bool,
        additionalHeaders: [String: String] = [:]
    ) async throws -> Data {
        guard let baseURL = baseURL else {
            throw SDKError.network(.serviceNotAvailable, "HTTP service not configured")
        }

        let url = buildURL(base: baseURL, path: path)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = payload

        return try await executeRequest(request, requiresAuth: requiresAuth, additionalHeaders: additionalHeaders)
    }

    /// GET request with raw response
    public func getRaw(
        _ path: String,
        requiresAuth: Bool
    ) async throws -> Data {
        guard let baseURL = baseURL else {
            throw SDKError.network(.serviceNotAvailable, "HTTP service not configured")
        }

        let url = buildURL(base: baseURL, path: path)
        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        return try await executeRequest(request, requiresAuth: requiresAuth)
    }

    // MARK: - Convenience Methods

    /// POST with JSON string body
    public func post(_ path: String, json: String, requiresAuth: Bool = false) async throws -> Data {
        guard let data = json.data(using: .utf8) else {
            throw SDKError.general(.validationFailed, "Invalid JSON string")
        }
        return try await postRaw(path, data, requiresAuth: requiresAuth)
    }

    /// POST with Encodable payload
    public func post<T: Encodable>(_ path: String, payload: T, requiresAuth: Bool = true) async throws -> Data {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(payload)
        return try await postRaw(path, data, requiresAuth: requiresAuth)
    }

    /// DELETE request
    public func delete(_ path: String, requiresAuth: Bool = true) async throws -> Data {
        guard let baseURL = baseURL else {
            throw SDKError.network(.serviceNotAvailable, "HTTP service not configured")
        }

        let url = buildURL(base: baseURL, path: path)
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"

        return try await executeRequest(request, requiresAuth: requiresAuth)
    }

    /// PUT request
    public func put(_ path: String, _ payload: Data, requiresAuth: Bool = true) async throws -> Data {
        guard let baseURL = baseURL else {
            throw SDKError.network(.serviceNotAvailable, "HTTP service not configured")
        }

        let url = buildURL(base: baseURL, path: path)
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.httpBody = payload

        return try await executeRequest(request, requiresAuth: requiresAuth)
    }

    // MARK: - Private Implementation

    private func buildURL(base: URL, path: String) -> URL {
        // Handle paths that start with "/" vs full URLs
        if path.hasPrefix("http://") || path.hasPrefix("https://") {
            return URL(string: path) ?? base.appendingPathComponent(path)
        }

        // Check if path contains query parameters
        if path.contains("?") {
            // Split path and query parameters
            let components = path.split(separator: "?", maxSplits: 1)
            let pathPart = String(components[0])
            let queryPart = String(components[1])

            // Build URL with query parameters using URLComponents
            guard var urlComponents = URLComponents(url: base, resolvingAgainstBaseURL: true) else {
                return base.appendingPathComponent(path)
            }
            let existingPath = urlComponents.path
            urlComponents.path = existingPath + pathPart
            urlComponents.query = queryPart

            return urlComponents.url ?? base.appendingPathComponent(path)
        }

        return base.appendingPathComponent(path)
    }

    private func executeRequest(_ request: URLRequest, requiresAuth: Bool, additionalHeaders: [String: String] = [:]) async throws -> Data {
        var request = request

        // Add additional headers if provided
        for (key, value) in additionalHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }

        // Add authorization header
        let token = try await resolveToken(requiresAuth: requiresAuth)
        if !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        // Execute request
        let (data, response) = try await session.data(for: request)

        // Validate response
        guard let httpResponse = response as? HTTPURLResponse else {
            throw SDKError.network(.invalidResponse, "Invalid HTTP response")
        }

        // Check status code
        // For device registration, 409 (Conflict) means device already exists, which is fine
        let isDeviceRegistration = request.url?.absoluteString.contains(RAC_ENDPOINT_DEV_DEVICE_REGISTER) ?? false
        let isSuccess = (200...299).contains(httpResponse.statusCode) || (isDeviceRegistration && httpResponse.statusCode == 409)

        guard isSuccess else {
            let error = parseHTTPError(
                statusCode: httpResponse.statusCode,
                data: data,
                url: request.url?.absoluteString ?? "unknown"
            )
            logger.error("HTTP \(httpResponse.statusCode): \(request.url?.absoluteString ?? "unknown")")
            throw error
        }

        // Log 409 as info for device registration (device already exists)
        if isDeviceRegistration && httpResponse.statusCode == 409 {
            logger.info("Device already registered (409) - treating as success")
        }

        return data
    }

    private func resolveToken(requiresAuth: Bool) async throws -> String {
        if requiresAuth {
            // Get token from C++ state, refreshing if needed
            if let token = CppBridge.State.accessToken, !CppBridge.State.tokenNeedsRefresh {
                return token
            }
            // Try refresh if we have refresh token
            if CppBridge.State.refreshToken != nil {
                try await CppBridge.Auth.refreshToken()
                if let token = CppBridge.State.accessToken {
                    return token
                }
            }
            // Fallback to API key if no OAuth token available
            // This supports API key-only authentication for production mode
            if let key = apiKey, !key.isEmpty {
                return key
            }
            throw SDKError.authentication(.authenticationFailed, "No valid authentication token")
        }
        // Use API key for non-auth requests
        return apiKey ?? ""
    }

    private func parseHTTPError(statusCode: Int, data: Data, url _: String) -> SDKError {
        // Try to parse error message from response body
        var errorMessage = "HTTP error \(statusCode)"

        // JSONSerialization returns heterogeneous dictionary for parsing unknown JSON error responses
        // swiftlint:disable:next avoid_any_type
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let message = json["message"] as? String {
                errorMessage = message
            } else if let error = json["error"] as? String {
                errorMessage = error
            } else if let hint = json["hint"] as? String {
                errorMessage = "\(errorMessage): \(hint)"
            }
        }

        switch statusCode {
        case 400:
            return SDKError.network(.httpError, "Bad request: \(errorMessage)")
        case 401:
            return SDKError.authentication(.authenticationFailed, errorMessage)
        case 403:
            return SDKError.authentication(.forbidden, errorMessage)
        case 404:
            return SDKError.network(.httpError, "Not found: \(errorMessage)")
        case 429:
            return SDKError.network(.httpError, "Rate limited: \(errorMessage)")
        case 500...599:
            return SDKError.network(.serverError, "Server error (\(statusCode)): \(errorMessage)")
        default:
            return SDKError.network(.httpError, "HTTP \(statusCode): \(errorMessage)")
        }
    }
}
